service/history/workflow/mutable_state_impl.go

```diff
@@ -6927,21 +6927,46 @@ func (ms *MutableStateImpl) applyUpdatesToStateMachineNodes(
 				// we don't have enough information to recreate all parents
 				return err
 			}
-			newNode, err := parentNode.AddChild(incomingPath[len(incomingPath)-1], nodeMutation.Data)
-			if err != nil {
-				return err
+
+			key := incomingPath[len(incomingPath)-1]
+			parentInternalNode := parentNode.InternalRepr()
+			machines, ok := parentInternalNode.Children[key.Type]
+			if ok {
+				if _, ok = machines.MachinesById[key.ID]; ok {
+					if ok {
+						return fmt.Errorf("%w: %v", hsm.ErrStateMachineAlreadyExists, key)
+					}
+				}
+			}
+
+			internalNode := &persistencespb.StateMachineNode{
+				Children:                      make(map[string]*persistencespb.StateMachineMap),
+				Data:                          nodeMutation.Data,
+				InitialVersionedTransition:    nodeMutation.InitialVersionedTransition,
+				LastUpdateVersionedTransition: nodeMutation.LastUpdateVersionedTransition,
+				TransitionCount:               1,
+			}
+			children, ok := parentInternalNode.Children[key.Type]
+			if !ok {
+				children = &persistencespb.StateMachineMap{MachinesById: make(map[string]*persistencespb.StateMachineNode)}
+				// Children may be nil if the map was empty and the proto message we serialized and deserialized.
+				if parentInternalNode.Children == nil {
+					parentInternalNode.Children = make(map[string]*persistencespb.StateMachineMap, 1)
+				}
+				parentInternalNode.Children[key.Type] = children
 			}
-			internalNode = newNode.InternalRepr()
+			children.MachinesById[key.ID] = internalNode
 		} else {
-			internalNode = currentNode.InternalRepr()
+			internalNode = node.InternalRepr()
 			if CompareVersionedTransition(nodeMutation.LastUpdateVersionedTransition, internalNode.LastUpdateVersionedTransition) == 0 {
 				continue
 			}
+			internalNode.Data = nodeMutation.Data
+			internalNode.InitialVersionedTransition = nodeMutation.InitialVersionedTransition
+			internalNode.LastUpdateVersionedTransition = nodeMutation.LastUpdateVersionedTransition
+			internalNode.TransitionCount++
+			node.InvalidateCache()
 		}
-		internalNode.Data = nodeMutation.Data
-		internalNode.InitialVersionedTransition = nodeMutation.InitialVersionedTransition
-		internalNode.LastUpdateVersionedTransition = nodeMutation.LastUpdateVersionedTransition
-		internalNode.TransitionCount++
```

Comment:
I would just keep this and assign internalNode in the `if` branch above. It'll shorten this function and remove some duplicated code.